File: models/crm_dashboard.py
from odoo import fields, models, api
from odoo.tools import date_utils


class SalesTeamInherit(models.Model):
    _inherit = "crm.team"

    crm_state_id = fields.Many2one("crm.stage")


class SaleInherit(models.Model):
    _inherit = "sale.order"

    def action_confirm(self):
        super(SaleInherit, self).action_confirm()
        self.team_id.crm_state_id = self.opportunity_id.stage_id


class CrmLead(models.Model):
    _inherit = "crm.lead"

    crm_manager_id = fields.Many2one("res.users", string="Crm Manager")

    @api.model
    def crm_lead(self):
        """lead,opportunity,expected revenue,revenue,win ratio card values"""
        record = {}
        lead_count = self.env["crm.lead"].search_count([('type', '=', 'lead'), ('user_id', '=', self.env.user.id)])
        opportunity_count = self.env["crm.lead"].search_count(
            [('type', '=', 'opportunity'), ('user_id', '=', self.env.user.id)])
        expected_revenue = sum(
            self.env["crm.lead"].search([('user_id', '=', self.env.user.id)]).mapped('expected_revenue'))
        revenue = round(sum(self.env["sale.order"].search([('user_id', '=', self.env.user.id)]).mapped("amount_total")),
                        2)
        win_count = (self.env["crm.lead"]).search_count(
            [('stage_id.is_won', '=', True), ('user_id', '=', self.env.user.id)])
        win_ratio = round((win_count / (lead_count + opportunity_count)) * 100, 2)

        record.update({
            "lead": lead_count,
            "opportunity": opportunity_count,
            "expected_revenue": expected_revenue,
            "revenue": revenue,
            "win_ratio": win_ratio
        })

        return record

    @api.model
    def crm_year(self):
        rec = {}
        lead_count = self.search([('type', '=', 'lead'), ('user_id', '=', self.env.user.id)])
        year_lead_count = len(
            [record.create_date.year for record in lead_count if fields.Date.today().year == record.create_date.year])

        opportunity_count = self.env["crm.lead"].search(
            [('type', '=', 'opportunity'), ('user_id', '=', self.env.user.id)])
        year_opp_count = len(
            [record.create_date.year for record in opportunity_count if
             fields.Date.today().year == record.create_date.year])

        year_expected_revenue = sum(self.env["crm.lead"].search([('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.year == fields.Date.today().year).mapped("expected_revenue"))
        year_revenue = round(sum(self.env["sale.order"].search([('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.year == fields.Date.today().year).mapped("amount_total")),
                             2)
        year_win_count = (self.env["crm.lead"]).search(
            [('stage_id.is_won', '=', True), ('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.year == fields.Date.today().year)
        year_win_list = len([i.id for i in year_win_count])
        year_win_ratio = round((year_win_list / (year_lead_count + year_opp_count)) * 100, 2)
        year_win_manager = (self.env["crm.lead"]).search([('stage_id.is_won', '=', True)]).filtered(
            lambda x: x.create_date.year == fields.Date.today().year)
        year_win_manager_count = len([i.id for i in year_win_manager])
        year_loss_manager = (self.env["crm.lead"]).search(
            [('active', '=', False)]).filtered(
            lambda x: x.create_date.year == fields.Date.today().year)
        year_loss_manager_count = len([i.id for i in year_loss_manager])

        rec.update({
            'lead_year': year_lead_count,
            'opportunity_year': year_opp_count,
            'expected_rev_year': year_expected_revenue,
            'year_revenue': year_revenue,
            'year_win_ratio': year_win_ratio,
            "year_win_manager_count": year_win_manager_count,
            "year_loss_manager_count": year_loss_manager_count

        })

        return rec

    @api.model
    def crm_month(self):
        rec = {}
        month_lead = self.env["crm.lead"].search(
            [('type', '=', 'lead'), ('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.month == fields.Date.today().month)
        month_lead_count = len([i.id for i in month_lead])

        month_opportunity = self.env["crm.lead"].search(
            [('type', '=', 'opportunity'), ('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.month == fields.Date.today().month)
        month_opp_count = len([i.id for i in month_opportunity])

        month_expected_revenue = sum(self.env["crm.lead"].search([('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.month == fields.Date.today().month).mapped("expected_revenue"))
        month_revenue = round(sum(self.env["sale.order"].search([('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.month == fields.Date.today().month).mapped("amount_total")),
                              2)
        month_win_count = (self.env["crm.lead"]).search(
            [('stage_id.is_won', '=', True), ('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.month == fields.Date.today().month)
        month_win_list = len([i.id for i in month_win_count])
        month_win_ratio = round((month_win_list / (month_lead_count + month_opp_count)) * 100, 2)
        month_win_manager = (self.env["crm.lead"]).search([('stage_id.is_won', '=', True)]).filtered(
            lambda x: x.create_date.month == fields.Date.today().month)
        month_win_manager_count = len([i.id for i in month_win_manager])
        month_loss_manager = (self.env["crm.lead"]).search(
            [('active', '=', False)]).filtered(
            lambda x: x.create_date.month == fields.Date.today().month)
        month_loss_manager_count = len([i.id for i in month_loss_manager])

        rec.update({
            "month_lead_count": month_lead_count,
            "month_opp_count": month_opp_count,
            "month_expected_revenue": month_expected_revenue,
            "month_revenue": month_revenue,
            "month_win_ratio": month_win_ratio,
            "month_win_manager_count": month_win_manager_count,
            "month_loss_manager_count": month_loss_manager_count,

        })

        return rec

    @api.model
    def crm_week(self):
        rec = {}
        week_lead = self.env["crm.lead"].search(
            [('type', '=', 'lead'), ('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.isocalendar()[1] == fields.Date.today().isocalendar()[1])
        week_lead_count = len([i.id for i in week_lead])

        week_opportunity = self.env["crm.lead"].search(
            [('type', '=', 'opportunity'), ('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.isocalendar()[1] == fields.Date.today().isocalendar()[1])
        week_opp_count = len([i.id for i in week_opportunity])

        week_expected_revenue = sum(self.env["crm.lead"].search([('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.isocalendar()[1] == fields.Date.today().isocalendar()[1]).mapped(
            "expected_revenue"))
        week_revenue = round(sum(self.env["sale.order"].search([('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.isocalendar()[1] == fields.Date.today().isocalendar()[1]).mapped("amount_total")),
                             2)
        week_win_count = (self.env["crm.lead"]).search(
            [('stage_id.is_won', '=', True), ('user_id', '=', self.env.user.id)]).filtered(
            lambda x: x.create_date.isocalendar()[1] == fields.Date.today().isocalendar()[1])
        week_win_list = len([i.id for i in week_win_count])
        week_win_ratio = round((week_win_list / (week_lead_count + week_opp_count)) * 100, 2)

        week_win_manager = (self.env["crm.lead"]).search([('stage_id.is_won', '=', True)]).filtered(
            lambda x: x.create_date.isocalendar()[1] == fields.Date.today().isocalendar()[1])
        week_win_manager_count = len([i.id for i in week_win_manager])
        week_loss_manager = (self.env["crm.lead"]).search(
            [('active', '=', False)]).filtered(
            lambda x: x.create_date.isocalendar()[1] == fields.Date.today().isocalendar()[1])
        week_loss_manager_count = len([i.id for i in week_loss_manager])

        rec.update({
            "week_lead_count": week_lead_count,
            "week_opp_count": week_opp_count,
            "week_expected_revenue": week_expected_revenue,
            "week_revenue": week_revenue,
            "week_win_ratio": week_win_ratio,
            "week_win_manager_count": week_win_manager_count,
            "week_loss_manager_count": week_loss_manager_count

        })
        return rec

    @api.model
    def crm_quarter(self):
        start_date, end_date = date_utils.get_quarter(fields.Date.today())
        rec = {}
        quarter_lead_count = self.env["crm.lead"].search_count(
            [('type', '=', 'lead'), ('user_id', '=', self.env.user.id), ('create_date', '>=', start_date),
             ('create_date', '<=', end_date)])

        quarter_opp_count = self.env["crm.lead"].search_count(
            [('type', '=', 'opportunity'), ('user_id', '=', self.env.user.id), ('create_date', '>=', start_date),
             ('create_date', '<=', end_date)])

        quarter_expected_revenue = sum(
            self.env["crm.lead"].search([('user_id', '=', self.env.user.id), ('create_date', '>=', start_date),
                                         ('create_date', '<=', end_date)]).mapped("expected_revenue"))

        quarter_revenue = round(
            sum(self.env["sale.order"].search([('user_id', '=', self.env.user.id), ('create_date', '>=', start_date),
                                               ('create_date', '<=', end_date)]).mapped("amount_total")),
            2)

        quarter_win_count = (self.env["crm.lead"]).search_count(
            [('stage_id.is_won', '=', True), ('user_id', '=', self.env.user.id), ('create_date', '>=', start_date),
             ('create_date', '<=', end_date)])
        quarter_win_ratio = round((quarter_win_count / (quarter_lead_count + quarter_opp_count)) * 100, 2)

        quarter_win_manager_count = (self.env["crm.lead"]).search_count(
            [('stage_id.is_won', '=', True), ('create_date', '>=', start_date),
             ('create_date', '<=', end_date)])

        quarter_loss_manager_count = (self.env["crm.lead"]).search_count(
            [('active', '=', False), ('create_date', '>=', start_date),
             ('create_date', '<=', end_date)])

        rec.update({
            "quarter_lead_count": quarter_lead_count,
            "quarter_opp_count": quarter_opp_count,
            "quarter_expected_revenue": quarter_expected_revenue,
            "quarter_revenue": quarter_revenue,
            "quarter_win_ratio": quarter_win_ratio,
            "quarter_win_manager_count": quarter_win_manager_count,
            "quarter_loss_manager_count": quarter_loss_manager_count
        })

        return rec

    @api.model
    def get_year_crm_activity(self):
        mail = self.env["mail.activity"].search([('res_model', '=', 'crm.lead')]).filtered(
            lambda x: x.create_date.year == fields.Date.today().year)
        mail_type_list = [re.activity_type_id.id for re in mail]
        year_crm_activity = {rec.name: mail_type_list.count(rec.id) for rec in mail.activity_type_id}

        return year_crm_activity

    @api.model
    def get_month_crm_activity(self):
        mail = self.env["mail.activity"].search([('res_model', '=', 'crm.lead')]).filtered(
            lambda x: x.create_date.month == fields.Date.today().month)
        mail_type_list = [re.activity_type_id.id for re in mail]
        month_crm_activity = {rec.name: mail_type_list.count(rec.id) for rec in mail.activity_type_id}

        return month_crm_activity

    @api.model
    def get_week_crm_activity(self):
        mail = self.env["mail.activity"].search([('res_model', '=', 'crm.lead')]).filtered(
            lambda x: x.create_date.isocalendar()[1] == fields.Date.today().isocalendar()[1])
        mail_type_list = [re.activity_type_id.id for re in mail]
        week_crm_activity = {rec.name: mail_type_list.count(rec.id) for rec in mail.activity_type_id}

        return week_crm_activity
    @api.model
    def get_quarter_crm_activity(self):
        start_date, end_date = date_utils.get_quarter(fields.Date.today())
        mail = self.env["mail.activity"].search([('res_model', '=', 'crm.lead'), ('create_date', '>=', start_date),
                                                 ('create_date', '<=', end_date)])
        mail_type_list = [re.activity_type_id.id for re in mail]
        quarter_crm_activity = {rec.name: mail_type_list.count(rec.id) for rec in mail.activity_type_id}

        return quarter_crm_activity

File: static/src/js/crm_dashboard.js
odoo.define('crm_dashboard.', function (require) {
    'use strict';
    var AbstractAction = require('web.AbstractAction');
    var core = require('web.core');
    var rpc = require('web.rpc');
    var QWeb = core.qweb;
    var session = require('web.session')
    var year_win_loss=[]
    var month_win_loss=[]
    var week_win_loss=[]
    var quarter_win_loss=[]
    var CrmDashboard = AbstractAction.extend({
        template: 'DashboardCrm',
        events: {
            'click .my_lead': 'my_lead',
            'click .opportunity': 'opportunity',
            'change #income_expense_values': function (e) {
                e.stopPropagation();
                var $target = $(e.target);
                var value = $target.val();
                if (value == "this_year") {
                    this.onclick_this_year($target.val());

                }
                else if (value == "this_quarter") {
                    this.onclick_this_quarter($target.val());
                }
                else if (value == "this_month") {
                    this.onclick_this_month($target.val());
                }
                else if (value == "this_week") {
                    this.onclick_this_week($target.val());
                }
            },
        },
        init: function (parent, action) {
            this._super(parent, action);
        },
        start: function () {

            var self = this;
            this._super().then(function () {

            }),


                rpc.query({
                    model: 'crm.lead',
                    method: 'crm_lead',
                    args: [],
                }).then(function (result) {
                    $('#my_leads').append('<span>' + result.lead + '</span>');
                    $('#my_opportunities').append('<span>' + result.opportunity + '</span>');
                    $('#expected_revenue').append('<span>' + result.expected_revenue + '</span>');
                    $('#revenue').append('<span>' + result.revenue + '</span>');
                    $('#win_ratio').append('<span>' + result.win_ratio + '</span>');

                })

        },


        onclick_this_year: function (ev) {
            var self = this;


            rpc.query({
                model: 'crm.lead',
                method: 'crm_year',
                args: [],
            })
                .then(function (result) {
                    $('#my_leads').empty()
                    $('#my_opportunities').empty()
                    $('#expected_revenue').empty()
                    $('#revenue').empty()
                    $('#win_ratio').empty()

                    $('#my_leads').append('<span>' + result.lead_year + '</span>');
                    $('#my_opportunities').append('<span>' + result.opportunity_year + '</span>');
                    $('#expected_revenue').append('<span>' + result.expected_rev_year + '</span>');
                    $('#revenue').append('<span>' + result.year_revenue + '</span>');
                    $('#win_ratio').append('<span>' + result.year_win_ratio + '</span>');

                    $("#weekChart").hide()
                    $('#monthChart').hide()
                    $('#quarterChart').hide()
                    $('#yearChart').show();

                    $('#year_crm_activity').show()
                    $('#month_crm_activity').hide()
                    $('#quarter_crm_activity').hide()
                    $('#week_crm_activity').hide()

                    year_win_loss.push(result.year_win_manager_count,result.year_loss_manager_count)
                    self.render_year_won_loss();
                    self.render_year_crm_activity();





                })
        },

        onclick_this_month: function (ev) {

            var self = this;
            rpc.query({
                model: 'crm.lead',
                method: 'crm_month',
                args: [],
            })
                .then(function (result) {
                    $('#my_leads').empty()
                    $('#my_opportunities').empty()
                    $('#expected_revenue').empty()
                    $('#revenue').empty()
                    $('#win_ratio').empty()

                    $('#my_leads').append('<span>' + result.month_lead_count + '</span>');
                    $('#my_opportunities').append('<span>' + result.month_opp_count + '</span>');
                    $('#expected_revenue').append('<span>' + result.month_expected_revenue + '</span>');
                    $('#revenue').append('<span>' + result.month_revenue + '</span>');
                    $('#win_ratio').append('<span>' + result.month_win_ratio + '</span>');

                    $("#weekChart").hide()
                    $('#quarterChart').hide()
                    $('#yearChart').hide()
                    $('#monthChart').show()

                    $('#year_crm_activity').hide()
                    $('#week_crm_activity').hide()
                    $('#quarter_crm_activity').hide()
                    $('#month_crm_activity').show()

                    month_win_loss.push(result.month_win_manager_count,result.month_loss_manager_count)
                    self.render_month_won_loss()
                    self.render_month_crm_activity();






                })
        },

        onclick_this_week: function (ev) {
            var self = this;
            rpc.query({
                model: 'crm.lead',
                method: 'crm_week',
                args: [],
            })
                .then(function (result) {
                    $('#my_leads').empty()
                    $('#my_opportunities').empty()
                    $('#expected_revenue').empty()
                    $('#revenue').empty()
                    $('#win_ratio').empty()

                    $('#my_leads').append('<span>' + result.week_lead_count + '</span>');
                    $('#my_opportunities').append('<span>' + result.week_opp_count + '</span>');
                    $('#expected_revenue').append('<span>' + result.week_expected_revenue + '</span>');
                    $('#revenue').append('<span>' + result.week_revenue + '</span>');
                    $('#win_ratio').append('<span>' + result.week_win_ratio + '</span>');


                    $('#yearChart').hide()
                    $('#monthChart').hide()
                    $('#quarterChart').hide()
                    $("#weekChart").show()

                    $('#year_crm_activity').hide()
                    $('#month_crm_activity').hide()
                    $('#quarter_crm_activity').hide()
                    $('#week_crm_activity').show()
                    week_win_loss.push(result.week_win_manager_count,result.week_loss_manager_count)
                    self.render_week_won_loss()
                    self.render_week_crm_activity()


                })
        },

        onclick_this_quarter: function (ev) {
            var self = this;
            rpc.query({
                model: 'crm.lead',
                method: 'crm_quarter',
                args: [],
            })
                .then(function (result) {
                    $('#my_leads').empty()
                    $('#my_opportunities').empty()
                    $('#expected_revenue').empty()
                    $('#revenue').empty()
                    $('#win_ratio').empty()

                    $('#my_leads').append('<span>' + result.quarter_lead_count + '</span>');
                    $('#my_opportunities').append('<span>' + result.quarter_opp_count + '</span>');
                    $('#expected_revenue').append('<span>' + result.quarter_expected_revenue + '</span>');
                    $('#revenue').append('<span>' + result.quarter_revenue + '</span>');
                    $('#win_ratio').append('<span>' + result.quarter_win_ratio + '</span>');

                    $('#yearChart').hide()
                    $('#monthChart').hide()
                    $("#weekChart").hide()
                    $('#quarterChart').show()

                    $('#year_crm_activity').hide()
                    $('#month_crm_activity').hide()
                    $('#quarter_crm_activity').show()
                    $('#week_crm_activity').hide()


                    quarter_win_loss.push(result.quarter_win_manager_count,result.quarter_loss_manager_count)
                    self.render_quarter_won_loss()
                    self.render_quarter_crm_activity()
                })
        },






        render_year_won_loss: function () {

            const ctx = $('#yearChart')
            new Chart(ctx, {
                type: 'bar',
                data: {
                    labels: ['Won', 'Loss'],
                    datasets: [{
                        label: '',
                        data: year_win_loss,
                        borderWidth: 1
                    }]
                },
                options: {
                    scales: {
                        y: {
                            beginAtZero: true
                        }
                    }
                }
            });
        },

        render_month_won_loss: function () {

            const ctx = $('#monthChart')
            new Chart(ctx, {
                type: 'bar',
                data: {
                    labels: ['Won', 'Loss'],
                    datasets: [{
                        label: '',
                        data: month_win_loss,
                        borderWidth: 1
                    }]
                },
                options: {
                    scales: {
                        y: {
                            beginAtZero: true
                        }
                    }
                }
            });
        },

    render_week_won_loss: function () {
            const ctx = $('#weekChart')
            new Chart(ctx, {
                type: 'bar',
                data: {
                    labels: ['Won', 'Loss'],
                    datasets: [{
                        label: '',
                        data: week_win_loss,
                        borderWidth: 1
                    }]
                },
                options: {
                    scales: {
                        y: {
                            beginAtZero: true
                        }
                    }
                }
            });
        },

        render_quarter_won_loss: function () {
            const ctx = $('#quarterChart')
            new Chart(ctx, {
                type: 'bar',
                data: {
                    labels: ['Won', 'Loss'],
                    datasets: [{
                        label: '',
                        data: quarter_win_loss,
                        borderWidth: 1
                    }]
                },
                options: {
                    scales: {
                        y: {
                            beginAtZero: true
                        }
                    }
                }
            });
        },


        render_year_crm_activity: function () {
            rpc.query({
                model: "crm.lead",
                method: "get_year_crm_activity",
            }).then(function (result) {
                var label = []
                var data1 = []
                $.each(result, function (index, name) {
                    label.push(index)
                    data1.push(name)
                })
                const ctx = $('#year_crm_activity')
                new Chart(ctx, {

                    type: 'bar',

                    data: {
                        labels: label,
                        datasets: [{
                            label: '',
                            data: data1,
                            borderWidth: 1
                        }]
                    },
                    options: {
                        scales: {
                            y: {
                                beginAtZero: true
                            }
                        }
                    }
                });
            });
        },

        render_month_crm_activity: function () {
            rpc.query({
                model: "crm.lead",
                method: "get_month_crm_activity",
            }).then(function (result) {
                var label = []
                var data1 = []
                $.each(result, function (index, name) {
                    label.push(index)
                    data1.push(name)
                })
                const ctx = $('#month_crm_activity')
                new Chart(ctx, {

                    type: 'bar',

                    data: {
                        labels: label,
                        datasets: [{
                            label: '',
                            data: data1,
                            borderWidth: 1
                        }]
                    },
                    options: {
                        scales: {
                            y: {
                                beginAtZero: true
                            }
                        }
                    }
                });
            });
        },

        render_week_crm_activity: function () {
            rpc.query({
                model: "crm.lead",
                method: "get_week_crm_activity",
            }).then(function (result) {
                var label = []
                var data1 = []
                $.each(result, function (index, name) {
                    label.push(index)
                    data1.push(name)
                })
                const ctx = $('#week_crm_activity')
                new Chart(ctx, {

                    type: 'bar',

                    data: {
                        labels: label,
                        datasets: [{
                            label: '',
                            data: data1,
                            borderWidth: 1
                        }]
                    },
                    options: {
                        scales: {
                            y: {
                                beginAtZero: true
                            }
                        }
                    }
                });
            });
        },

        render_quarter_crm_activity: function () {
            rpc.query({
                model: "crm.lead",
                method: "get_quarter_crm_activity",
            }).then(function (result) {
                var label = []
                var data1 = []
                $.each(result, function (index, name) {
                    label.push(index)
                    data1.push(name)
                })
                const ctx = $('#quarter_crm_activity')
                new Chart(ctx, {

                    type: 'bar',

                    data: {
                        labels: label,
                        datasets: [{
                            label: '',
                            data: data1,
                            borderWidth: 1
                        }]
                    },
                    options: {
                        scales: {
                            y: {
                                beginAtZero: true
                            }
                        }
                    }
                });
            });
        },







        my_lead: function (e) {
            this.do_action({
                name: "My Leads",
                type: 'ir.actions.act_window',
                res_model: 'crm.lead',
                view_mode: 'tree,form,calendar',
                views: [[false, 'list'], [false, 'form']],
                domain: [['user_id', '=', session.uid], ['type', '=', 'lead']],
                target: 'current',
            })
        },

        opportunity: function (e) {
            this.do_action({
                name: 'Opportunity',
                type: 'ir.actions.act_window',
                res_model: 'crm.lead',
                view_mode: 'tree,form',
                views: [[false, 'list'], [false, 'form']],
                domain: [['user_id', '=', session.uid], ['type', '=', 'opportunity']],
                target: 'current',
            })
        },








    });
    core.action_registry.add("crm_dashboard", CrmDashboard);
    return CrmDashboard;
});